### DISCUSSION TITLE
Use of subgraph isomorphism in this project?

### DISCUSSION THREAD

**Discussion by j2kun** [TuGraph-family/tugraph-db#417] [Q&A] [answered]
Hi there,

I'm doing background research for a book on graph algorithms in production software systems. I noticed this project has a subgraph isomorphism routine, but it's not clear to me where it is called, or what exactly it is used for. I imagine it is used to query the database for graphs that contain an isomorphic subgraph, but I'd like to know why that is useful for the clients of the database library. For example, in relation to the use cases listed at https://www.tugraph.org/case

**Comment by qishipengqsp** [comment, marked as answer]
Dear Jeremy, I'm delighted to hear about your interest in TuGraph!

To give you an example, let's consider financial applications where isomorphic subgraph matching, commonly referred to as `pattern matching`, plays a critical role. In the context of financial systems, vertices symbolize financial entities, while edges denote the interactions between them. Utilizing pattern matching queries on such graphs enables engineers to detect and identify specific patterns indicating potential fraudulent activities, such as money laundering or account theft. This method forms the cornerstone of our risk management framework within graph computing. The same logic can be applied to energy systems.

For further details, please examine the [LDBC FinBench](https://arxiv.org/abs/2306.15975) which we proposed to assess the performance of graph databases in financial scenario. This benchmark highlights some intriguing patterns, much like the ones previously mentioned.

Regarding the database library, we offer it to our clients so they can run their algorithms autonomously, without relying on the TuGraph database services.

Looking forward to our continued discussion!

**Comment by j2kun** [comment]
>  detect and identify specific patterns indicating potential fraudulent activities

What I'm curious about is what sort of graph-theoretical properties would correspond to fraud in this context?

**Comment by qishipengqsp** [comment]
The identification of fraudulent activities is contingent upon the underlying graph schema. Below are key patterns that may suggest fraudulent behavior:

- Transfer Loops/Cycles: Large transfer loops or cycles within a graph, where V denotes accounts and E signifies fund transfers between these accounts, could be indicative of money laundering.
- Fusiform Patterns: These structures suggest a distinct type of illicit fund transfer activity. The pattern typically involves a source account A sending funds to a destination account C through a series of intermediary accounts (B1, B2, B3).
- Label-Enhanced Analysis: By tagging accounts with labels, such as denoting them as malicious, further analysis becomes possible. This includes tracking funds transferred from flagged accounts and identifying accounts accessed by compromised devices.
- Advanced Graph Algorithms: More complex patterns may require graph-level algorithms for detection. Examples include employing community detection algorithms to uncover criminal networks or using path-finding algorithms for fund tracing.

These examples illustrate that when a graph schema is thoughtfully designed, it can effectively encapsulate behavioral patterns, allowing fraudulent activities to be interpreted and analyzed through graph-based representations.

**Comment by qishipengqsp** [comment]
@j2kun This discussion is inactive for one month. Looking forward to your book if you find it helpful. Please feel free to continue this discussion if I did not make it clear for you.
